### DISCUSSION TITLE
VT100 - RS323 - Websocket - Beginner questions

### DISCUSSION THREAD

**Discussion by JohannesNeu** [sourcelair/xterm.js#4883] [Q&A] [answered]
Hi Everyone,

I am absolutely new to xterm.js and frontend development in general, so I appreciate any links or help.

My current setup consists of a computer with an RS323 Interface that sends all boot/terminal messages to the serial output.
An attached MAX3232 turns them into UART, which is then read by micropython on an ESP32 and forwarded to a websocket.

I can print the messages line by line on a website as follows:
> <<< None
<<< b'[ 0.000000] Command line: BOOT_IMAGE=/vmlinuz-6.2.0-36-generic root=/dev/'
<<< b'[ 0.000000] KERNEL supported cpus:\r\n'
<<< None
<<< b'[ 0.000000] AMD AuthenticAMD\r\n'
<<< b'[ 0.000000] Centau'
<<< b'[ 0.000000] zhaoxin Shanghai \r\n'
<<< b'[ 0.000000] BIOS-e820: [mem 0x0000000000000000-0x000000000009ffff] usable\r\n'
<<< None
<<< b'[ 0.000000] BIOS-e820: [mem 0x0000000000100000-0x0000000009d0efff] usable\r\n'
<<< b'[ 0.000000] BIOS-e820: [mem 0x000000000a000000-0x000000000'
<<< b'[ 0.000000] BIOS-e820: [mem 0x000000000a200000-0x000000000a210fff] ACPI NVS\r\n'
<<< None
<<< b'[ 0.000000] BIOS-e820: [mem 0x000000000b000000-0x000000000b01f'
<<< b'[ 0.000000] BIOS-e820: [mem 0x000000000b020000-0x00000000c3246fff] usable\r\n'

To make all this a bit more interactive and use it as an IPMI replacement I would like to provide xterm.js via the micropython webserver (microdot) to my browser and connect xterm.js with the websocket.

My problem: I have no idea how to "provide" xterm.js to my client computer, all I could find so far were install tutorials that used npm. My questions would be:

1. What files would I need to copy to the ESP32 webserver to allow xterm.js to be loaded on the client PC?
2. What would the websocket need to output/accept to work with xterm.js? (Currently it is only a line by line output)

Thanks for your help.

**Comment by jerch** [comment, marked as answer]
I think the most basic document setup would look  like this (untested):
```html
<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <title>terminal test</title>
    <link rel="stylesheet" href="node_modules/xterm/css/xterm.css">
    <script src="node_modules/xterm/lib/xterm.js"></script>
    <script src="node_modules/xterm-addon-attach/lib/xterm-addon-attach.js"></script>
  </head>
  <body> 
    <div id="terminal"></div>
    <script>
      const term = new Terminal(yourConfigOptions);
      const attachAddon = new AttachAddon.AttachAddon(yourWebSocket);
      term.loadAddon(attachAddon);
      term.open(document.getElementById('terminal'));
    </script>
  </body>
</html>
```
so the needed files to be served by your webserver are `xterm.js` & `xterm.css` from the xterm package, and `xterm-addon-attach.js` from the xterm-addon-attach package. You can either extract those files manually from the packages, or just use npm to install them in a stub project and use the paths as shown above.
There are more things to consider, like your webserver must support websockets somehow, otherwise this setup wont work, or security and flow control handling. In general you should have a good understanding of HTTP mechs or this might end up as a security nightmare.
Last but not least the terminal's functionality can be further enhanced by more addons - whether you should use those, depends on concrete cicumstances (e.g. the fit addon might not be a good idea here, as boot prompts often have a fixed idea of the screen size).

**Comment by JohannesNeu** [comment]
Hi @jerch 
thank you very much for your detailled answer, with it I was able to set up a rudimentary display!

I absolutely agree with you on the security aspect. Next step would probably be some more work on the websocket format:
![grafik](https://github.com/xtermjs/xterm.js/assets/39328632/de4820ee-3a3e-4066-bc1d-555bf1b90020)

Looking at this, would you happen to know what format xterm.js would expect from the websocket?
At the moment I am reading line by line from UART and turn it into a string, which seems a  bit odd.

**Comment by JohannesNeu** [comment]
Simply forwarding the bytes to xterm.js and replacing all Nones with "" was enough. It looks really great!
I will need to look into customizing and  the security aspects and enabling sending messages, but I am very happy now:

![grafik](https://github.com/xtermjs/xterm.js/assets/39328632/60404c1d-8109-445e-b62c-e53ee2215914)

**Reply by jerch** [reply]
Yupp, directly forwarding bytes should work as long as the text encoding is UTF8.

**Reply by jerch** [reply]
One additional note - since you forward `bytes` directly you should also encode the `None` as `b''` instead of just `''`. Idk which websocket lib you use on python side and whether it exposes the byte vs string difference, but in general the websocket protocol treats byte and string content differently, which might create issues on terminal side. You should always stick to sending either bytes or string content, never mixing them.

**Reply by JohannesNeu** [reply]
A very good point, I changed it to `b''` thanks.
